zdsp: Add ARC DSPLIB backend for zdsp

SiyuanCheng-CN opened the pull request:
Introduce ARC DSPLIB backend zdsp library for ARC target.
Add __agu to map with ARC DSPLIB
Add basicmath tests for ARC backend

stephanosio commented on subsys/dsp/Kconfig:
```diff
@@ -32,6 +38,16 @@ config DSP_BACKEND_CUSTOM
 	  Rely on the application to provide a custom DSP backend. The implementation should be
 	  added to the 'zdsp' build target by the application or one of its modules.
 
+config DSP_BACKEND_ARC
+	bool "Use the mwdt library as the math backend"
+	depends on ARCMWDT_LIBC
+	depends on CMSIS_DSP
+	select DSP_BACKEND_HAS_STATIC
+	select DSP_POINTER_HAS_RESTRICT
+	help
+	  Implement the various zephyr DSP functions using the MWDT-DSP library. This feature
+	  requires the MetaWare toolchain and CMSIS module to be selected.
```
Why does `DSP_BACKEND_ARC` depend on `CMSIS_DSP`?

My understanding was that Synopsys had their own ARC DSP libraryloosely based on CMSIS-DSP, is that not the case?

SiyuanCheng-CN replied:
Synopsys DSP library currently doesn't have corresponding functions like clip(), and(), or(), not(), xor() comparing to CMSIS DSP library. These functions are mapped to CMSIS library temporarily.

yperess commented on subsys/dsp/Kconfig:
```diff
@@ -13,10 +13,16 @@ if DSP
 	prompt "DSP library backend selection"
 	default DSP_BACKEND_CMSIS if CMSIS_DSP
 	default DSP_BACKEND_CUSTOM
+	default DSP_BACKEND_ARC
```
Since there is no `if` above this, this default will never trigger. I think you should move the default up so it looks like:

```
default DSP_BACKEND_ARC if ARC
default DSP_BACKEND_CUSTOM
```

evgeniy-paltsev replied:
I guess it's even need to be

```
default DSP_BACKEND_ARC if ARC && "$(ZEPHYR_TOOLCHAIN_VARIANT)" = "arcmwdt"
default DSP_BACKEND_CUSTOM
```

as this is dsp backend from arc mwdt toolchain.

yperess commented on include/zephyr/dsp/dsp.h:
```diff
@@ -17,6 +17,20 @@
+#define DATA_ATTR
+#endif
+
+#ifdef CONFIG_DSP_POINTER_HAS_RESTRICT
+#define RESTRICT __restrict
+#else
+#define RESTRICT
+#endif
```
I don't believe that we have a resolution on https://github.com/zephyrproject-rtos/zephyr/discussions/52617 yet. Can this PR be done without `__restrict` for now until we can discuss further?

yperess reviewed (changes requested):
Please remove the use of `__restrict` as we cannot guarantee that `a += b` will be logically equivalent to `a = a + b`. We need to come to a consensus if we want to:
1. always use `__restrict`
2. never use `__restrict`
3. Allow `__restrict` based on Kconfig (knowing that any common code using the DSP will have to be implemented in 2 different ways).

My vote is to never use `__restrict` as it complicates things too much, see https://github.com/zephyrproject-rtos/zephyr/discussions/52617

ruuddw commented:
@yperess Did you see my earlier comments in see #52617 ?
I think the discussion on using __restrict or not is not the correct one, it should be on allowing in-place processing or not (a += b). That should be defined as part of the DSP API semantics.

An implementation with __restrict can still fail on 'a += b' and vice versa (we checked the ARC DSP lib, also with __restrict optimizations in-place processing functions fine).

yperess commented:
> @yperess Did you see my earlier comments in see #52617 ? I think the discussion on using __restrict or not is not the correct one, it should be on allowing in-place processing or not (a += b). That should be defined as part of the DSP API semantics.
> 
> An implementation with __restrict can still fail on 'a += b' and vice versa (we checked the ARC DSP lib, also with __restrict optimizations in-place processing functions fine).

Sure, we can frame the question around "should we support in-place operation" but I really see it as just the same question as should we use `__restrict`. I understand that it currently works, I think it's important to recognize that the ISO C standard defines in-place operations with `__restrict` as undefined. So my concern is an unknown version of an unknown compiler building Zephyr might break. An interesting study https://www.semanticscholar.org/paper/Why-Programmer-specified-Aliasing-is-a-Bad-Idea-Mock/ad21ead2416ccacc8a0bb84bcb93e66f5a6da5e0 (though I can only access the summary and the PDF seems to be missing on this site and others). I'm contacting our sensor team at Android to double check what their stance is though the more I read about this, the more I'm convinced that we should avoid `__restrict`.

ruuddw commented:
The __restrict is/should be optional, and only selected if the implementation is known to be compliant with the chosen API semantics. We can add some extra tests to check if in-place semantics are honored, that would be good to have for non-restrict implementations as well. For the basic math it may be trivial that in-place works, but there is no guarantee by not using __restrict.

yperess commented:
> The __restrict is/should be optional, and only selected if the implementation is known to be compliant with the chosen API semantics. We can add some extra tests to check if in-place semantics are honored, that would be good to have for non-restrict implementations as well. For the basic math it may be trivial that in-place works, but there is no guarantee by not using __restrict.

Right, but the issue is that if I write a subsystem (which I'm actively working on) for sensors that does some calibration and virtual sensors that heavily leverage the DSP subsystem I need to know that it'll work on all targets. So my concern is that if I have a need for something like `a += b` in my sensor subsystem. Do I need to have 2 implementations that fork on `CONFIG_DSP_USES_RESTRICTED`? That seems to not scale.

Here's my proposal:
1. Lets get this backend merged without `__restricted`
2. Lets create another PR or issue for `__restricted`. I strongly feel that we cannot use a Kconfig as we're basically saying that the API contract is different depending on the configuration. So the discussion should be around do we want to allow in-place memory operations or not.

I've yet to find a real example that demonstrates a big performance improvement for `__restricted` so I would love to see more data bout that in the next PR/issue. Everything I've found so far says it provides marginal performance improvements.

ruuddw commented:
The indication I got from our DSP experts is that the performance impact of __restrict will depend mostly on the depth of processor pipeline: for ARC EM class processors probably small, for ARC HS and VPX big.
ARC DSP library uses __restrict in the headers/public API, without the __restrict option for the Zephyr API it cannot be used.
As said before, let's decouple the use of __restrict or other pointer attributes from the API semantics.

SiyuanCheng-CN commented:
Removed __restrict related content now, consider adding it next time.

ruuddw commented:
@yperess, @stephanosio : I believe Siyuan addressed your comments, can you please re-review? Thanks.

SiyuanCheng-CN commented:
@yperess @stephanosio: Added in-place operation in every test to check if in-place processing is functioning correctly.

yperess commented on include/zephyr/dsp/dsp.h:
```diff
@@ -17,6 +17,14 @@
 #endif
 
+#ifdef CONFIG_DSP_BACKEND_ARC
+#define DSP_DATA __agu
```
I think that as we get more and more architecture backends we'll get more of these attributes. Can we use the same pattern as the static and have this as a Kconfig without a prompt in case ? maybe `config DSP_BACKEND_HAS_AGU`? Similarly for the data attribute, we can have a string Kconfig for the section. My goal is to make as few custom (per architecture) switches.

yperess replied:
I'm still a bit lost about the scalability of this approach, can you explain in more details how you see this when the next attribute comes up?

So lets assume we have `CONFIG_DSP_BACKEND_HAS_AGU` and a new attribute `CONFIG_DSP_BACKEND_HAS_X`, does the code now look like:

```
#if defined(CONFIG_DSP_BACKEND_HAS_AGU) && defined(CONFIG_DSP_BACKEND_HAS_X)
#define DSP_DATA __agu __x
#elif defined(CONFIG_DSP_BACKEND_HAS_AGU) && !defined(CONFIG_DSP_BACKEND_HAS_X)
#define DSP_DATA __agu
#elif !defined(CONFIG_DSP_BACKEND_HAS_AGU) && defined(CONFIG_DSP_BACKEND_HAS_X)
#define DSP_DATA __x
#endif
```

ruuddw replied:
I'd expect these would be mutual exclusive with at most a single attribute at a time. The _HAS refers to a specific hardware feature or flavor of DSP acceleration, so I don't expect a combination of _HAS_AGU and _HAS_X likely.
But now I understand your scalability concern better.  Can we address it in the future when the case of HAS_AGU && HAS_X occurs?

SiyuanCheng-CN replied:
I also expect these are mutual exclusive, so the code shouldn't be such complex. I tried to use string Kconfig to directly pass attributes to C macro `DSP_DATA` and `DSP_STATIC_DATA` so that there is no if statement in code. Kconfig and C macro definition might look like:
```
config DSP_DATA
	string
	default "__agu" if DSP_BACKEND_ARCMWDT
	default "__x" if DSP_BACKEND_X
```
`#define DSP_DATA CONFIG_DSP_DATA`
But I found double quotes are introduced to C macro by make itself, which is unrecognizable by compiler. If this can be solved, the code can be simple.

yperess commented on tests/subsys/dsp/basicmath/CMakeLists.txt:
```diff
@@ -14,3 +14,8 @@ target_sources(app PRIVATE
 target_include_directories(app PRIVATE ${ZEPHYR_BASE}/tests/lib/cmsis_dsp)
+
+if(COMPILER STREQUAL arcmwdt)
+get_property(Z_ARC_DSP_OPTIONS GLOBAL PROPERTY z_arc_dsp_options)
+target_compile_options(app PRIVATE ${Z_ARC_DSP_OPTIONS})
```
nit: indentation

yperess commented on subsys/dsp/arc/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+
+zephyr_include_directories(public)
+
+zephyr_include_directories_ifdef(CONFIG_DSP_BACKEND_ARC
+      ${ARCMWDT_TOOLCHAIN_PATH}/MetaWare/arc/lib/src/dsp/include/
+    )
```
nit: indentation, I believe should be:
```
zephyr_include_directories_ifdef(CONFIG_DSP_BACKEND_ARC
    ${ARCMWDT_TOOLCHAIN_PATH}/MetaWare/arc/lib/src/dsp/include/
)
```

yperess reviewed (changes requested):
Please also update the documentation at https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/services/dsp/index.rst

stephanosio commented on include/zephyr/dsp/dsp.h:
```diff
@@ -17,6 +17,18 @@
+#define DSP_DATA
+#endif
+
+#ifdef CONFIG_DSP_BACKEND_HAS_XDATA_SECTION
+#define DATA_ATTR __attribute__((section(".Xdata")))
+#else
+#define DATA_ATTR
+#endif
```
`DATA_ATTR` has multiple problems:

1. It is too generic -- if any, it should be called `DSP_DATA_ATTR`.
2. It is in conflict with `DSP_DATA` -- note that both `DSP_DATA` and `DATA_ATTR` are "data attributes."

In general, I would say `DATA_ATTR` is a misnomer and should be reworked as `#define DSP_STATIC_DATA DSP_DATA __attribute__((section(".Xdata")))`.

stephanosio commented on tests/subsys/dsp/basicmath/src/f16.c:
```diff
@@ -30,6 +30,21 @@ static void test_zdsp_add_f16(
 	/* Run test function */
 	zdsp_add_f16((float16_t *)input1, (float16_t *)input2, output, length);
 
+	/* Validate output */
+	zassert_true(
+		test_snr_error_f16(length, output, (float16_t *)ref,
+			SNR_ERROR_THRESH),
+		ASSERT_MSG_SNR_LIMIT_EXCEED);
+
+	zassert_true(
+		test_rel_error_f16(length, output, (float16_t *)ref,
+			REL_ERROR_THRESH),
+		ASSERT_MSG_REL_ERROR_LIMIT_EXCEED);
+
+	/* Validate in-place processing */
+	memcpy(output, (float16_t *)input1, length * sizeof(float16_t));
+	zdsp_add_f16(output, (float16_t *)input2, output, length);
```
Please make this a separate test (i.e. `test_zdsp_add_f16_in_place`) since this is testing a special type of operation.

stephanosio commented on tests/subsys/dsp/basicmath/src/f32.c:
```diff
@@ -219,30 +256,27 @@ DEFINE_TEST_VARIANT4(basic_math_f32, zdsp_scale_f32, 0p5_11, in_com1, 0.5f, ref_
 
 	/* Allocate output buffer */
-	output = malloc(1 * sizeof(float32_t));
+	output = (DSP_DATA float32_t *)malloc(1 * sizeof(float32_t));
```
Do we really need the attributes specified by the `DSP_DATA` for the type cast here?

SiyuanCheng-CN replied:
It's needed for mwdt compiler.

stephanosio commented on tests/subsys/dsp/basicmath/src/f32.c:
```diff
@@ -219,30 +256,27 @@ DEFINE_TEST_VARIANT4(basic_math_f32, zdsp_scale_f32, 0p5_11, in_com1, 0.5f, ref_
-		test_rel_error_f32(1, output, (float32_t *)ref,
-			REL_ERROR_THRESH),
-		ASSERT_MSG_REL_ERROR_LIMIT_EXCEED);
+	/* No meaningless in-place processing validation for dot production */
```
This comment is really unnecessary.

stephanosio commented on tests/subsys/dsp/basicmath/src/f32.pat:
```diff
@@ -1,4 +1,4 @@
-static const uint32_t in_com1[256] = {
+static DSP_DATA DATA_ATTR const uint32_t in_com1[256] = {
```
With `DSP_STATIC_DATA` as proposed above, this can be changed to:

```suggestion
static DSP_STATIC_DATA const uint32_t in_com1[256] = {
```

stephanosio commented on subsys/dsp/Kconfig:
```diff
@@ -32,6 +39,17 @@ config DSP_BACKEND_CUSTOM
 	  Rely on the application to provide a custom DSP backend. The implementation should be
 	  added to the 'zdsp' build target by the application or one of its modules.
 
+config DSP_BACKEND_ARC
```
`DSP_BACKEND_ARC` sounds rather vague to me given that this backend is for the MWDT DSP library and can only be used with the MetaWare toolchain.

Something like `DSP_BACKEND_ARCMWDT` or `DSP_BACKEND_MWDTDSP` would be more adequate.

stephanosio reviewed (changes requested):
Please split these changes into more manageable commits.

For example, the changes adding "in-place operation" tests should be in a separate commit from the changes adding the ARC DSPLIB backend.

MaureenHelm commented:
> Please split these changes into more manageable commits.
> 
> For example, the changes adding "in-place operation" tests should be in a separate commit from the changes adding the ARC DSPLIB backend.

I agree. It would be nice to see the DSP_DATA changes made in a separate commit before introducing the ARC-specific support. There are also a lot of whitespace changes that should be made in a separate commit so we can isolate functional changes from cosmetic ones.

Commit:
zdsp: Introduce attribute marco for third-party backend

Introduce DSP_DATA and DSP_STATIC_DATA attribute marco
to support third-party backend

Signed-off-by: Siyuan Cheng <siyuanc@synopsys.com>

Commit:
zdsp: add ARC DSPLIB backend for zdsp

Introduce ARC DSPLIB backend zdsp library for ARC target.
Add agu and restrict attributes to map with ARC DSPLIB

Signed-off-by: Siyuan Cheng <siyuanc@synopsys.com>

Commit:
zdsp: add in-place operaton test for basicmath

Add in-place calculatoin test for every functions in basicmath

Signed-off-by: Siyuan Cheng <siyuanc@synopsys.com>

SiyuanCheng-CN commented:
@MaureenHelm Thanks for the review. Already made `DSP_DATA` and `DSP_STATIC_DATA` changes into a separate commit.

yperess reviewed (changes requested):
Not really requesting a change, but I do want to get clarity about https://github.com/zephyrproject-rtos/zephyr/pull/52478/files#r1053641805

ruuddw commented:
@yperess, @stephanosio: I believe all your earlier concerns have been addressed, ok to get this included in this release?

stephanosio commented on subsys/dsp/arc/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2022 Synopsys
```
Since the backend is now called DSP_BACKEND_**ARCMWDT**, please update the relevant file paths to `subsys/dsp/arcmwdt/`.

stephanosio commented on tests/subsys/dsp/basicmath/testcase.yaml:
```diff
@@ -19,3 +19,8 @@ tests:
       - CONFIG_FPU=y
     min_flash: 128
     min_ram: 64
+  zdsp.basicmath.arc:
```
same here

```suggestion
  zdsp.basicmath.arcmwdt:
```

stephanosio commented on doc/services/dsp/index.rst:
```diff
@@ -50,7 +50,9 @@ added to `subsys/dsp/Kconfig`_ along with the required dependencies and the
-linked in at `subsys/dsp/CMakeLists.txt`_.
+linked in at `subsys/dsp/CMakeLists.txt`_. To add architecture-specific attributes,
+its corresponding Kconfig option should be added to `subsys/dsp/Kconfig`_ and use
+them to update ``DSP_DATA`` and ``DSP_STATIC_DATA`` in `include/zephyr/dsp/dsp.h`_.
```
To fix the documentation build failure:

```suggestion
them to update ``DSP_DATA`` and ``DSP_STATIC_DATA`` in :file:`include/zephyr/dsp/dsp.h`.
```

p.s. the rest of this doc should be using `:file:` as well.

stephanosio reviewed (commented):
Looks ok to me. Just some final comments.

abrodkin commented:
@stephanosio unfortunately @SiyuanCheng-CN is away this week due to a New Year holiday in China so very unlikely we'll be able to address your comments this week so that the PR is merged before the upcoming feature freeze (I guess it happens later this week). I'm wondering if there's a chance to get it merged as an exclusion (given its scope and current state of the review) a bit later once @SiyuanCheng-CN is back next week?

yperess commented on subsys/dsp/Kconfig:
```diff
@@ -32,6 +39,17 @@ config DSP_BACKEND_CUSTOM
+config DSP_BACKEND_ARCMWDT
+	bool "Use the mwdt library as the math backend"
+	depends on ARCMWDT_LIBC
+	depends on CMSIS_DSP
```
I just noticed you have this `depends on CMSIS_DSP` here, on line 24 we select the CMSIS backend if we have the `CMSIS_DSP` selected. so that means that this backend will never be selected by default. Do we really need this dependency?

SiyuanCheng-CN replied:
Synopsys DSP library currently doesn't have corresponding functions like clip(), and(), or(), not(), xor() comparing to CMSIS DSP library. These functions are mapped to CMSIS library temporarily. I think it is needed to inform users to arcmwdt backend is partly dependent on cmsis_dsp.

stephanosio commented:
> @stephanosio unfortunately @SiyuanCheng-CN is away this week due to a New Year holiday in China so very unlikely we'll be able to address your comments this week so that the PR is merged before the upcoming feature freeze (I guess it happens later this week). I'm wondering if there's a chance to get it merged as an exclusion (given its scope and current state of the review) a bit later once @SiyuanCheng-CN is back next week?

@abrodkin Since this PR is fairly self-contained, I have no objection to making this an exception; but, that will need to be approved by the TSC as per the documented process. I will briefly mention this during the TSC meeting this week; but, I cannot make any guarantees on whether an exception will be granted.

abrodkin commented:
@stephanosio there's no pressure for this PR and we may easily get it into the next release, it was more of an attempt to work-around a bit unfortunate timeline.

fabiobaltieri commented:
@SiyuanCheng-CN can you rebase please?

carlescufi commented:
@SiyuanCheng-CN please rebase and solve the conflict

SiyuanCheng-CN commented:
@fabiobaltieri @carlescufi Already rebased, thx

yperess commented on subsys/dsp/Kconfig:
```diff
@@ -13,9 +13,16 @@ if DSP
 choice DSP_BACKEND
 	prompt "DSP library backend selection"
 	default DSP_BACKEND_CMSIS if CMSIS_DSP
+	default DSP_BACKEND_ARCMWDT if ARC && "$(ZEPHYR_TOOLCHAIN_VARIANT)" = "arcmwdt"
```
I would suggest moving this up one line so we choose the right optimized default (feel free to do this in another PR in order to avoid getting another round of reviews).
